### PR TITLE
Bug 820446 - issue with bugzilla and persona login

### PR DESCRIPTION
**Pull request by jedp**
Monday 2012-12-10 while rolling out train-2012.11.23 to production there was
an issue with login to bugzilla with persona.

jrgm theorized that this was a string/int comparison error
(http://irclog.gr/#show/irc.mozilla.org/identity/187967)

git bisecting confirms that he is correct, and that the problem crept in here:
https://github.com/mozilla/browserid/commit/d494afbbce176697998f3b39764200d8732ad256#L28L77

### CODE REVIEW

File: lib/verifier/certassertion.js [hunk 1]
```diff
@@ -35,7 +35,8 @@ logger.debug("This verifier will accept assertions issued by " + HOSTNAME);
 //   it might be strangely formed.
 function compareAudiences(want, got) {
   function normalizeParsedURL(u) {
-    if (!u.port) u.port = u.protocol === 'https:' ? 443 : 80;
+    // return string because url.parse returns string
+    if (!u.port) u.port = u.protocol === 'https:' ? '443' : '80';
```
Comment (ozten) [inline lib/verifier/certassertion.js:39]: > url.parse returns string

Ya, that is a bit odd.

Another way to roll, would be `parseInt`, but your solution is solid.

File: lib/verifier/certassertion.js [hunk 2]
```diff
@@ -155,4 +156,5 @@ function verify(assertion, audience, successCB, errorCB) {
     }, errorCB);
 }
 
+exports.compareAudiences = compareAudiences;
```
Comment (ozten) [inline lib/verifier/certassertion.js:159]: Thanks for making that unit testable.

File: tests/verifier-test.js [hunk 3]
```diff
@@ -34,6 +36,36 @@ const TEST_EMAIL = 'someuser@somedomain.com',
...
+  'audiences': {
+    'http://fakesite.com and http://fakesite.com:80': matchesAudience(true),
+    'https://fakesite.com and https://fakesite.com:443': matchesAudience(true),
+    'http://fakesite.com:8000 and http://fakesite.com:8000': matchesAudience(true),
+    'https://fakesite.com:9000 and https://fakesite.com:9000': matchesAudience(true),
+
+    'http://fakesite.com:8100 and http://fakesite.com:80': matchesAudience(false),
+    'https://fakesite.com:9100 and https://fakesite.com:443': matchesAudience(false),
+    'http://fakesite.com:80 and http://fakesite.com:8000': matchesAudience(false),
+    'https://fakesite.com:443 and https://fakesite.com:9000': matchesAudience(false)
```
Comment (ozten) [inline tests/verifier-test.js:64]: Very cool.

**Comment by ozten** [comment]
Travis went to crazy town on the MYSQL tests. I'll run locally.
https://travis-ci.org/mozilla/browserid/jobs/3622187

**Comment by ozten** [comment]
Passes test locally. I also ran the tests against the old code and verified the new tests fail:

```
audiences http://fakesite.com and http://fakesite.com:80
✗ should match
  » true == false

audiences https://fakesite.com and https://fakesite.com:443
✗ should match
  » true == false
✗ Broken » 85 honored ∙ 2 broken (10.819s)
```

**Comment by jrgm** [comment]
"""Travis went to crazy town on the MYSQL tests. I'll run locally.
https://travis-ci.org/mozilla/browserid/jobs/3622187"""

Cool new feature: load that link, sign in with github and if you have commit privs, switch to the build tab, click the control widget and then 'restart build'. It will try again.

**Pull request merged**